File: src/main/java/de/eugenbernwald/fretboardtrainer/service/ToneEvaluator.java
package de.eugenbernwald.fretboardtrainer.service;

import be.tarsos.dsp.AudioEvent;
import be.tarsos.dsp.pitch.PitchDetectionHandler;
import be.tarsos.dsp.pitch.PitchDetectionResult;
import de.eugenbernwald.fretboardtrainer.model.Tone;
import javafx.animation.AnimationTimer;
import javafx.beans.property.DoubleProperty;
import javafx.beans.property.SimpleDoubleProperty;

public class ToneEvaluator extends AnimationTimer implements PitchDetectionHandler {

    private float currentPitch;

    private Tone tone;
    private DoubleProperty progress;
    private CorrectToneListener  correctToneListener;

    public ToneEvaluator(Tone tone, CorrectToneListener listener) {
        this.currentPitch = 0.0f;
        this.tone = tone;
        this.progress = new SimpleDoubleProperty(0.0);
        this.correctToneListener = listener;
    }

    public DoubleProperty getProgress() {
        return this.progress;
    }

    public Tone getTone() {
        return this.tone;
    }

    public void setTone(Tone tone) {
        this.tone = tone;
    }

    @Override
    public void handle(long now) {
        if(this.currentPitch < tone.getUpperLimit() && this.currentPitch > tone.getLowerLimit()){
            hit();
        }
        else {
            miss();
        }
    }

    @Override
    public void handlePitch(PitchDetectionResult pitchDetectionResult, AudioEvent audioEvent) {
        this.currentPitch = pitchDetectionResult.getPitch();
    }

    private void hit(){
        progress.setValue(progress.get() + (1.0/50));

        if(progress.get() >= 1.0){
            progress.setValue(0);
            correctToneListener.onCorrectTone();
        }

    }

    private void miss() {

        if(progress.get() > 0){
            progress.setValue(progress.get() - (1.0/50));
        }
        else {
            progress.setValue(0);
        }
    }
}

File: src/main/java/de/eugenbernwald/fretboardtrainer/service/ToneEvaluator.kt
package de.eugenbernwald.fretboardtrainer.service

import be.tarsos.dsp.AudioEvent
import be.tarsos.dsp.pitch.PitchDetectionHandler
import be.tarsos.dsp.pitch.PitchDetectionResult
import de.eugenbernwald.fretboardtrainer.model.Tone
import javafx.animation.AnimationTimer
import javafx.beans.property.DoubleProperty
import javafx.beans.property.SimpleDoubleProperty

class ToneEvaluator(var tone: Tone, private val correctToneListener: () -> Unit  ) : AnimationTimer(), PitchDetectionHandler {

    private var currentPitch: Float = 0.0f
    val progress: DoubleProperty

    init {
        this.currentPitch = 0.0f
        this.progress = SimpleDoubleProperty(0.0)
    }

    override fun handle(now: Long) {
        if (this.currentPitch < tone.upperLimit && this.currentPitch > tone.lowerLimit) {
            hit()
        } else {
            miss()
        }
    }

    override fun handlePitch(pitchDetectionResult: PitchDetectionResult, audioEvent: AudioEvent) {
        this.currentPitch = pitchDetectionResult.pitch
    }

    private fun hit() {
        progress.value = progress.get() + 1.0 / 50

        if (progress.get() >= 1.0) {
            progress.setValue(0)
            correctToneListener()
        }

    }

    private fun miss() {

        if (progress.get() > 0) {
            progress.value = progress.get() - 1.0 / 50
        } else {
            progress.setValue(0)
        }
    }
}
